### ISSUE TITLE
Loading Datetime Objects

### ISSUE THREAD

**Issue by HolgerPeters** [marshmallow-code/marshmallow#656]
Maybe this is not so much of an issue but a misunderstanding on my side, so I'd think the following pytest-style unit test should not raise a validation error, since a `datetime.datetime` instance should satisfy the property of being a a datetime already:

```
from marshmallow import Schema, fields
import datetime


class SomeSchema(Schema):
    element = fields.DateTime()


def test_datetime_loading_accepts_datetime_object():
    schema = SomeSchema(strict=True)
    res = schema.load({"element": datetime.datetime.now()})
```

I do however get a validation error `marshmallow.exceptions.ValidationError: {'element': ['Not a valid datetime.']}`

**Comment by RAlexeev** [comment]
Try changing `res = schema.load({"element": datetime.datetime.now()})` to `res = schema.load({"element": str(datetime.datetime.now())})`. It helps me.

**Comment by JulienPalard** [comment]
Hit the same "problem", rewording it by using the documentation example it's like:

```
from datetime import date
from marshmallow import Schema, fields


class ArtistSchema(Schema):
    name = fields.Str()


class AlbumSchema(Schema):
    title = fields.Str()
    release_date = fields.Date()
    artist = fields.Nested(ArtistSchema())


bowie = dict(name='David Bowie')
album = dict(artist=bowie, title='Hunky Dory', release_date=date(1971, 12, 17))

schema = AlbumSchema()
print(schema.validate(album))
```

It gives `{'release_date': ['Not a valid date.']}` which looks wrong to me, it *is* a valid date and it satisfies the schema so validation should pass.

I also think that for every valid object, its unserialized form should be valid too, not only for datetimes.

**Comment by JulienPalard** [comment]
I'm temporarily using this so deserialization accepts datetimes instances:

```
class MyDateTimeField(fields.DateTime):
    def _deserialize(self, value, attr, data):
        if isinstance(value, datetime):
            return value
        return super()._deserialize(value, attr, data)
```

**Comment by whoiswes** [comment]
I also ran into this, albeit in a slightly different way.  We were loading in RFC822 formatted strings from JSON, but dumping ISO8601.  This worked great until 2.13.5, after which we need to explicitly convert the RFC format to ISO format in pre-load.

Is there a marshmallow native way of loading one format but dumping another, specific to datetimes?  Or, as the comments above hint at, could/should DateTime load multiple user-definable input formats and dump 1?

Sorry if this is a dumb question or otherwise inappropriate, my python-fu is not quite at the level where I fully understand what's happening under the covers.

**Comment by clarkdowner** [comment]
I'm also running into issues serializing datetimes whether passed in as DateTime objects or as an ISO formatted string:
`MarshalResult(data={...}, errors={'birth_date': ['"1994-11-24T00:00:00Z" cannot be formatted as a datetime.']})`

Looking at the source code for the DateTime Field it appears to be expecting a string in the above format, no? Although I am new to Python so I could be misinterpreting the code.

**Comment by aklaver** [comment]
If are serializing(dump) a DateTime field then you need to provide a datetime object:

```
import datetime
from marshmallow import Schema, fields
from dateutil import tz
class UserSchema(Schema):
    name = fields.Str()
    email = fields.Email()
    created_at = fields.DateTime()
usr_sch = UserSchema()
usr = {'name': "Adrian Klaver", 'email': "aklaver@example.com", 'created_at': datetime.datetime(2017, 11, 22, 9, 58, tzinfo=tz.tzutc())}
usr_sch.dump(usr)
MarshalResult(data={'name': 'Adrian Klaver', 'created_at': '2017-11-22T09:58:00+00:00', 'email': 'aklaver@example.com'}, errors={})
```
If you are going the other way deserializing(load) then you can use a string:

```
usr = {'name': "Adrian Klaver", 'email': "aklaver@example.com", 'created_at': '2017-11-22T09:58:00Z'}
usr_sch.load(usr)
UnmarshalResult(data={'name': 'Adrian Klaver', 'created_at': datetime.datetime(2017, 11, 22, 9, 58, tzinfo=tzutc()), 'email': 'aklaver@example.com'}, errors={})


```

**Comment by clarkdowner** [comment]
That makes sense. Thank you @aklaver

**Comment by appcuarium** [comment]
Hi,

In the latest version, loading schema for a DateTime field behaves differently from version 3.0.0b2 when passing a Date string in format "YY-MM-DD".

Input:
`date_string = '2017-12-12'`

Output from 3.0.0b2:
`'2017-12-12T00:00:00Z'`

Output from 3.0.0b4:
`"Not a valid period of time."`.

The schema looks like this:
```
class ReportsDateRange(Schema):
    start = fields.DateTime()
    end = fields.DateTime()
    field = fields.Str()
```

Is this a bug or a feature? Thanks.

**Comment by aklaver** [comment]
I am going to say it is the below, so a feature:

3.0.0b3 (2017-08-20)
Bug fixes:
    Includes bug fixes from releases 2.13.5 and 2.13.6.

2.13.5 (2017-04-12)
Bug fixes:
     Fix validation of iso8601-formatted dates (#556). Thanks @lafrech for reporting.

https://github.com/marshmallow-code/marshmallow/issues/556

See comment from May 31.

**Comment by lafrech** [comment]
Indeed, the answer is that deserialization expects a datetime as string and validation fails if the input is a datetime object already.

@JulienPalard [proposed a way](https://github.com/marshmallow-code/marshmallow/issues/656#issuecomment-318587611) to subclass `DateTime` field to accept datetime objects and wrote:

> I also think that for every valid object, its unserialized form should be valid too, not only for datetimes.

Possibly. This is what umongo does [for datetimes](https://github.com/Scille/umongo/blob/master/umongo/fields.py#L166) ([umongo](https://github.com/Scille/umongo) is a MondoDB ODM based on marshmallow).

This, applying to all fields, not only `DateTime`, makes sense as a feature request. If interested, please open a new issue.
